compiler/src/iree/compiler/DispatchCreation/FormDispatchRegions.cpp

```diff
@@ -54,58 +54,252 @@ namespace mlir::iree_compiler::DispatchCreation {
+  Operation *rootOp;
+  // All operations to be fused with the root op. This does not include
+  // `rootOp`.
+  SetVector<Operation *> fusedOps;
+  DenseMap<Operation *, AffineMap> loopMaps;
+};
+} // namespace
+
+void FusionGroup::insert(Operation *op) {
+  assert(!fusedOps.contains(op) && "op already fused");
+  FailureOr<AffineMap> map = getRootParallelLoopToOpMap(op);
+  if (succeeded(map)) {
+    loopMaps.insert({op, map.value()});
+  } else {
+    // TODO(IanWood1): some ops can be fused but don't implement
+    // `LinalgFusionOpInterface` e.g. `tensor.insert_slice`. For now, use
+    // a null map. These ops are always at the edges of the dispatch which means
+    // they are the last to be added so a null map should be fine.
+    loopMaps.insert({op, AffineMap()});
+  }
+  fusedOps.insert(op);
 }
-/// Returns the fusion groups for the given `op`.
-static SmallVector<int64_t, 1> getFusionGroups(Operation *op) {
-  SmallVector<int64_t, 1> fusionGroups = {};
-  if (auto fusionGroupsAttr = op->getAttrOfType<ArrayAttr>(kFusionGroupsAttr)) {
-    fusionGroups = llvm::map_to_vector(fusionGroupsAttr, [](Attribute attr) {
-      return llvm::cast<IntegerAttr>(attr).getInt();
-    });
+
+FailureOr<AffineMap>
+FusionGroup::getRootParallelLoopToOpMap(Operation *op) const {
+  auto fusionOp = dyn_cast<IREE::LinalgExt::LinalgFusionOpInterface>(op);
+  if (!fusionOp) {
+    return failure();
   }
-  return fusionGroups;
-}
-/// Appends the given `op` to the `newGroups` fusion groups.
-static void appendToFusionGroup(Operation *op, ArrayRef<int64_t> newGroups) {
-  SmallVector<int64_t> fusionGroups = getFusionGroups(op);
-  fusionGroups.append(newGroups.begin(), newGroups.end());
-  op->setAttr(kFusionGroupsAttr, Builder(op).getI64ArrayAttr(fusionGroups));
-}
-/// Removes the fusion groups attribute.
-static void removeFusionGroupsAttribute(Operation *op) {
-  op->removeAttr(kFusionGroupsAttr);
+
+  bool isConsumer = llvm::any_of(
+      op->getOperands(), [&](Value v) { return contains(v.getDefiningOp()); });
+
+  AffineMap newMap;
+  if (isConsumer) {
+    for (OpOperand &operand : op->getOpOperands()) {
+      Operation *definingOp = operand.get().getDefiningOp();
+      if (!contains(definingOp)) {
+        continue;
+      }
+
+      auto fusionProducer =
+          operand.get()
+              .getDefiningOp<IREE::LinalgExt::LinalgFusionOpInterface>();
+      if (!fusionProducer) {
+        return failure();
+      }
+      auto it = loopMaps.find(fusionProducer);
+      assert(it != loopMaps.end());
+
+      AffineMap producerResultMap = fusionProducer.getIndexingMapMatchingResult(
+          cast<OpResult>(operand.get()));
+      AffineMap consumerOperandMap = fusionOp.getMatchingIndexingMap(&operand);
+      if (!producerResultMap || !consumerOperandMap ||
+          !producerResultMap.isProjectedPermutation() ||
+          !consumerOperandMap.isProjectedPermutation()) {
+        return failure();
+      }
+      AffineMap inverseMap =
```

Comment:
Might be worth adding some comments here of this form

1. producerMap is a map from (producerIterationSpace) -> (operand data space)
2. consumerMap is a map from (consumerIterationSpace) -> (operand data space)
3. inverseMap is a map from (operand data space) -> (producerIterationSpace)
4. composedMap is a map from (producerIterationSpace) -> (consumerIterationSpace)

So now I am not sure I follow what the `composedMap.compose(producerMap)` is doing